From e62885c67c2e6101383c06d6a7ceae20a74d5887 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marcin=20Wcis=C5=82o?= <marcin.wcislo@conclusive.pl>
Date: Tue, 8 Jul 2025 02:00:07 +0200
Subject: [PATCH] rcu: Upgrade rcu_swap_protected() to rcu_replace_pointer()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

jira VULN-7633
cve-pre CVE-2023-1281
commit-author Paul E. McKenney <paulmck@kernel.org>
commit a63fc6b75cca984c71f095282e0227a390ba88f3

Although the rcu_swap_protected() macro follows the example of
swap(), the interactions with RCU make its update of its argument
somewhat counter-intuitive.  This commit therefore introduces
an rcu_replace_pointer() that returns the old value of the RCU
pointer instead of doing the argument update.  Once all the uses of
rcu_swap_protected() are updated to instead use rcu_replace_pointer(),
rcu_swap_protected() will be removed.

Link: https://lore.kernel.org/lkml/CAHk-=wiAsJLw1egFEE=Z7-GGtM6wcvtyytXZA1+BHqta4gg6Hw@mail.gmail.com/
	Reported-by: Linus Torvalds <torvalds@linux-foundation.org>
[ paulmck: From rcu_replace() to rcu_replace_pointer() per Ingo Molnar. ]
	Signed-off-by: Paul E. McKenney <paulmck@kernel.org>
	Cc: Bart Van Assche <bart.vanassche@wdc.com>
	Cc: Christoph Hellwig <hch@lst.de>
	Cc: Hannes Reinecke <hare@suse.de>
	Cc: Johannes Thumshirn <jthumshirn@suse.de>
	Cc: Shane M Seymour <shane.seymour@hpe.com>
	Cc: Martin K. Petersen <martin.petersen@oracle.com>
(cherry picked from commit a63fc6b75cca984c71f095282e0227a390ba88f3)
	Signed-off-by: Marcin Wcisło <marcin.wcislo@conclusive.pl>
---
 include/linux/rcupdate.h | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/include/linux/rcupdate.h b/include/linux/rcupdate.h
index c783ecaf1458..6387b45ebfe4 100644
--- a/include/linux/rcupdate.h
+++ b/include/linux/rcupdate.h
@@ -969,6 +969,24 @@ static inline notrace void rcu_read_unlock_sched_notrace(void)
 #define rcu_assign_pointer(p, v) \
 	__rcu_assign_pointer((p), (v), __rcu)
 
+/**
+ * rcu_replace_pointer() - replace an RCU pointer, returning its old value
+ * @rcu_ptr: RCU pointer, whose old value is returned
+ * @ptr: regular pointer
+ * @c: the lockdep conditions under which the dereference will take place
+ *
+ * Perform a replacement, where @rcu_ptr is an RCU-annotated
+ * pointer and @c is the lockdep argument that is passed to the
+ * rcu_dereference_protected() call used to read that pointer.  The old
+ * value of @rcu_ptr is returned, and @rcu_ptr is set to @ptr.
+ */
+#define rcu_replace_pointer(rcu_ptr, ptr, c)				\
+({									\
+	typeof(ptr) __tmp = rcu_dereference_protected((rcu_ptr), (c));	\
+	rcu_assign_pointer((rcu_ptr), (ptr));				\
+	__tmp;								\
+})
+
 /**
  * RCU_INIT_POINTER() - initialize an RCU protected pointer
  *
-- 
2.47.3

